main.tf

```
@@ -0,0 +1,124 @@
+resource "aws_vpc" "webapp" {
+  cidr_block           = var.cidr
+  enable_dns_support   = var.enable_dns_support
+  enable_dns_hostnames = var.enable_dns_hostnames
+
+  tags = merge(
+    local.base_tags,
+    {
+      Name = "${local.name}-${local.environment}"
+    }
+  )
+}
+resource "aws_subnet" "public" {
+  for_each = toset(var.public_subnets)
+
+  cidr_block              = each.value
+  availability_zone       = element(data.aws_availability_zones.available.names, index(var.public_subnets, each.value))
+  vpc_id                  = aws_vpc.webapp.id
+  map_public_ip_on_launch = true
+
+  tags = merge(
+    local.base_tags,
+    {
+      Name = "public-${each.value}"
+    }
+  )
+}
+
+resource "aws_internet_gateway" "public" {
+  vpc_id = aws_vpc.webapp.id
+
+  tags = merge(
+    local.base_tags,
+    {
+      Name = "${local.name}-igw"
+    }
+  )
+}
+
+resource "aws_route_table" "public" {
+  vpc_id = aws_vpc.webapp.id
+
+  route {
+    cidr_block = "0.0.0.0/0"
+    gateway_id = aws_internet_gateway.public.id
+  }
+
+  tags = merge(
+    local.base_tags,
+    {
+      Name = "public-${aws_internet_gateway.public.id}"
+    }
+  )
+}
+
+resource "aws_route_table_association" "public" {
+  for_each = aws_subnet.public
+
+  subnet_id      = each.value.id
+  route_table_id = aws_route_table.public.id
+}
+
+resource "aws_subnet" "private" {
+  for_each = toset(var.private_subnets)
+
+  cidr_block        = each.value
+  availability_zone = element(data.aws_availability_zones.available.names, index(var.private_subnets, each.value))
+  vpc_id            = aws_vpc.webapp.id
+
+  tags = merge(
+    local.base_tags,
+    {
+      Name = "private-${each.value}"
+    }
+  )
+}
+
+resource "aws_eip" "private" {
+  for_each = aws_subnet.public
+  domain   = "vpc"
+  tags = merge(
+    local.base_tags,
+    {
+      Name = "${local.name}-${each.value.availability_zone}"
+    }
+  )
+}
+
+resource "aws_nat_gateway" "private" {
+  for_each = toset(var.public_subnets)
+
+  allocation_id = aws_eip.private[each.value].id
+  subnet_id     = aws_subnet.public[each.value].id # Assuming you're associating it with a public subnet
```

If looping directly over the public subnet resources, you can get this directly here now instead
```suggestion
  subnet_id     = each.value.id
```

---

Since we're looping over both the subnets and EIPs, using `aws_subnet.public` in for_each directly would break the correlation between subnet_id and allocation_id. We're naming the resources consistently by CIDR block, so using toset(var.private_subnets) ensures that each subnet is correctly paired with its corresponding EIP. 


aws_subnet.public also doesn't return it's CIDR as an argument to be reference after declaration.